### PR TITLE
Use gather send [10657]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR extends the transport interface to avoid copying user data when encapsulating the serialized payload into the RTPS datagram.

### CODE REVIEW

**Comment by EduPonz** [comment]
@richiprosima please test discovery-server

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp [hunk 1]
```diff
@@ -154,7 +154,7 @@ void test_UDPv4Transport::get_ips(
 }
 
 bool test_UDPv4Transport::send(
-        const octet* send_buffer,
+        const std::array<asio::const_buffer, 3>& send_buffer,
         uint32_t send_buffer_size,
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/test_UDPv4Transport.cpp:158]:
```suggestion
        size_t send_buffer_size,
```

Maybe we should unify the criteria in all the `send` methods. Changing the type to `size_t` where the parameter is a size.

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp [hunk 2]
```diff
@@ -192,21 +192,21 @@ bool test_UDPv4Transport::send(
 }
 
 bool test_UDPv4Transport::send(
-        const octet* send_buffer,
-        uint32_t send_buffer_size,
+        const std::array<asio::const_buffer, 3>& send_buffers,
+        uint32_t total_bytes,
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/test_UDPv4Transport.cpp:196]: I do not understand why in this case you have changed the variable name when the methods called within all refer to `send_buffer_size` or `size` (in the case of `log_drop`).

File: src/cpp/rtps/transport/test_UDPv4Transport.h [hunk 3]
```diff
@@ -101,15 +101,15 @@ class test_UDPv4Transport : public UDPv4Transport
             const fastrtps::rtps::octet* buffer,
             uint32_t size);
     bool packet_should_drop(
-            const fastrtps::rtps::octet* send_buffer,
-            uint32_t send_buffer_size);
+            const std::array<asio::const_buffer, 3>& send_buffers,
+            uint32_t total_bytes);
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/test_UDPv4Transport.h:105]: For example, here the name of the variable has been changed but this change has not been applied in the implementation.

File: src/cpp/rtps/transport/UDPSenderResource.hpp [hunk 4]
```diff
@@ -43,14 +43,26 @@ class UDPSenderResource : public fastrtps::rtps::SenderResource
...
+            uint32_t total_bytes,
             fastrtps::rtps::LocatorsIterator* destination_locators_begin,
             fastrtps::rtps::LocatorsIterator* destination_locators_end,
             const std::chrono::steady_clock::time_point& max_blocking_time_point) -> bool
                 {
-                    return transport.send(data, dataSize, socket_, destination_locators_begin,
+                    assert(num_buffers < 4);
+
+                    std::array<asio::const_buffer, 3> asio_buffers;
+                    uint32_t total_size = 0;
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/UDPSenderResource.hpp:57]:
```suggestion
                    size_t total_size = 0;
```

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Partial review until commit 261686e6849379531ccfb6c76ee79fabe7c89c97.

File: include/fastdds/rtps/messages/RTPSMessageSenderInterface.hpp [hunk 5]
```diff
@@ -77,28 +77,16 @@ class RTPSMessageSenderInterface
...
          * Send a message through this interface.
          *
-         * @param message Pointer to the buffer with the message already serialized.
+         * @param buffers Array of buffers to gather.
+         * @param num_buffers Number of elements on @c buffers.
+         * @param total_bytes Total size of the raw data. Should be equal to the sum of the @c length field of all buffers.
          * @param max_blocking_time_point Future timepoint where blocking send should end.
          */
         virtual bool send(
                 const NetworkBuffer* message,
```
Comment (JLBuenoLopez-eProsima) [inline include/fastdds/rtps/messages/RTPSMessageSenderInterface.hpp]: The doxygen parameter name should be the same as the one declared in the function.

```suggestion
                const NetworkBuffer* buffers,
```

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

**Comment by JLBuenoLopez-eProsima** [comment]
There are some tests that are failing consistently with this PR and not in master (running `--repeat-until-fail 100` in master succeeds):

* 242: BlackboxTests_FastRTPS.PubSubFragments.AsyncPubSubAsNonReliableVolatileData300kb.Transport
* 897: BlackboxTests_DDS_PIM.PubSubFragments.AsyncPubSubAsNonReliableVolatileData300kb.Transport (this one has failed in the Linux CI)
* 555: BlackboxTests_FastRTPS.BlackBox.TCP_TLS: the following error message appears:
```
[RTCP_MSG_IN Error] Bad RTCP header identifier, closing connection. -> Function Receive
```
* 1353: BlackboxTests_DDS_PIM.BlackBox.TCP_TLS: master sometimes blocks and timeouts (rarely). This also happens with this PR (frequently) but the following message is printed that in master does not appear:
```
[RTCP_MSG_IN Error] Bad RTCP header identifier, closing connection. -> Function Receive
```
* 1831: SHMTransportTests.buffer_recover: in master sometimes happens but locally in my machine is failing consistently with this PR:
```
1845: terminate called without an active exception
    Test #1845: SHMTransportTests.buffer_recover ...Child aborted***Exception:  12.57 sec
```

File: include/fastdds/rtps/network/NetworkBuffer.hpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file NetworkBuffer.hpp
+ */
+
+#ifndef FASTDDS_RTPS_NETWORK_NETWORKBUFFER_HPP
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/rtps/network/NetworkBuffer.hpp:19]: The proposal on [this PR](https://github.com/eProsima/cpp-style/pull/6) is to use both. Will change this one

File: src/cpp/rtps/transport/shared_mem/SharedMemTransport.h [hunk 7]
```diff
@@ -135,19 +135,19 @@ class SharedMemTransport : public TransportInterface
...
-     * It must not exceed the send_buffer_size fed to this class during construction.
-     * @param socket channel we're sending from.
-     * @param remote_locator Locator describing the remote destination we're sending to.
-     * @param only_multicast_purpose
-     * @param timeout Maximum time this function will block
+     * Blocking Send to a list of destinations.
+     * @param buffers Array of buffers to gather.
+     * @param num_buffers Number of elements on @c buffers.
+     * @param total_bytes Total size of the raw data. Should be equal to the sum of the @c length field of all
+     * buffers. It must not exceed the send_buffer_size fed to this class during construction.
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/shared_mem/SharedMemTransport.h:142]: It still applies. Method `shared_mem_segment_->alloc_buffer` will be called with `total_bytes`. The maximum allowed size for that call is `configuration_.segment_size()`, which returns the same value as `configuration_.min_send_buffer_size()`.

That said, I think it is better to change the comment to `segment_size`. Will do

File: src/cpp/rtps/transport/UDPSenderResource.hpp [hunk 8]
```diff
@@ -44,13 +44,25 @@ class UDPSenderResource : public fastrtps::rtps::SenderResource
...
-            uint32_t dataSize,
+            const NetworkBuffer* buffers,
+            size_t num_buffers,
+            uint32_t total_bytes,
             fastrtps::rtps::LocatorsIterator* destination_locators_begin,
             fastrtps::rtps::LocatorsIterator* destination_locators_end,
             const std::chrono::steady_clock::time_point& max_blocking_time_point) -> bool
                 {
-                    return transport.send(data, dataSize, socket_, destination_locators_begin,
+                    assert(num_buffers < 4);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/UDPSenderResource.hpp:54]: The check should be `<= 3`. What I have done is to create a `constexpr` on `SenderResource` indicating the maximum value that will be passed to `num_buffers`.

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/test_UDPv4Transport.cpp:158]: I agree with the last sentence

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/test_UDPv4Transport.cpp:196]: I have refactored log_drop

File: src/cpp/rtps/transport/UDPSenderResource.hpp (hunk 4 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/UDPSenderResource.hpp:57]: I have renamed this to `total_num_bytes`

File: src/cpp/rtps/messages/submessages/DataMsg.hpp [hunk 9]
```diff
@@ -206,9 +233,17 @@ bool RTPSMessageCreator::addSubmessageData(
...
+        }
+    }
+    else
     {
-        added_no_error &= CDRMessage::addOctet(msg, 0);
+        pending_padding = align;
+        msg->pos += align;
     }
 
     //if(align > 0)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/messages/submessages/DataMsg.hpp]: Fixed in 078753eb7c0435718764d59770b367084291e46c

File: include/fastdds/rtps/network/NetworkBuffer.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/rtps/network/NetworkBuffer.hpp]: Fixed in 8a8fa52ff44ee669001a209917548cf077b9f0ce

File: src/cpp/rtps/transport/shared_mem/SharedMemTransport.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/shared_mem/SharedMemTransport.h]: Fixed in c4629f4f7dccd61a0971561735e06dd8f9211e50

File: src/cpp/rtps/transport/UDPSenderResource.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/UDPSenderResource.hpp]: Fixed in 2d12790368ee29888049189dc3d749525a36200d

File: src/cpp/rtps/transport/TCPSenderResource.hpp [hunk 10]
```diff
@@ -42,13 +42,25 @@ class TCPSenderResource : public fastrtps::rtps::SenderResource
...
-            uint32_t dataSize,
+            const NetworkBuffer* buffers,
+            size_t num_buffers,
+            uint32_t total_bytes,
             fastrtps::rtps::LocatorsIterator* destination_locators_begin,
             fastrtps::rtps::LocatorsIterator* destination_locators_end,
             const std::chrono::steady_clock::time_point&) -> bool
                 {
-                    return transport.send(data, dataSize, channel_, destination_locators_begin,
+                    assert(num_buffers < 4);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/TCPSenderResource.hpp]: Fixed in 2d12790368ee29888049189dc3d749525a36200d

File: src/cpp/rtps/transport/TCPSenderResource.hpp [hunk 11]
```diff
@@ -42,13 +42,25 @@ class TCPSenderResource : public fastrtps::rtps::SenderResource
...
+
+                    std::array<asio::const_buffer, 3> asio_buffers;
+                    uint32_t total_size = 0;
+                    for (size_t i = 0; i < num_buffers; ++i)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/TCPSenderResource.hpp:56]: Fixed in 2d12790368ee29888049189dc3d749525a36200d

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/test_UDPv4Transport.cpp]: Changed in 33959d01d5afabd1b0a76f7df4238b632fca3b57

File: src/cpp/rtps/transport/test_UDPv4Transport.h (hunk 3 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/test_UDPv4Transport.h:105]: Changed in 33959d01d5afabd1b0a76f7df4238b632fca3b57

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/test_UDPv4Transport.cpp:196]: Refactor in e926974cab9c6b3a8597397828a84ba076c82c2d

File: src/cpp/rtps/transport/UDPSenderResource.hpp (hunk 4 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/UDPSenderResource.hpp]: See f5a9a8fc3b28de332619b6c2a40dbcbf48a7f4a7

File: src/cpp/rtps/transport/TCPSenderResource.hpp [hunk 12]
```diff
@@ -42,13 +42,25 @@ class TCPSenderResource : public fastrtps::rtps::SenderResource
...
+            uint32_t total_bytes,
             fastrtps::rtps::LocatorsIterator* destination_locators_begin,
             fastrtps::rtps::LocatorsIterator* destination_locators_end,
             const std::chrono::steady_clock::time_point&) -> bool
                 {
-                    return transport.send(data, dataSize, channel_, destination_locators_begin,
+                    assert(num_buffers < 4);
+
+                    std::array<asio::const_buffer, 3> asio_buffers;
+                    uint32_t total_size = 0;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/transport/TCPSenderResource.hpp]: See f5a9a8fc3b28de332619b6c2a40dbcbf48a7f4a7

**Comment by MiguelCompany** [comment]
@JLBuenoLopez-eProsima Doxygen changes are still pending, but the rest can be reviewed

File: src/cpp/rtps/transport/TCPChannelResourceBasic.cpp [hunk 13]
```diff
@@ -157,14 +156,17 @@ size_t TCPChannelResourceBasic::send(
     {
         if (header_size > 0)
         {
-            std::array<asio::const_buffer, 2> buffers;
+            std::array<asio::const_buffer, max_required_buffers + 1> buffers;
             buffers[0] = asio::buffer(header, header_size);
-            buffers[1] = asio::buffer(data, size);
+            for (size_t i = 0; i < 3; ++i)
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/TCPChannelResourceBasic.cpp:161]:
```suggestion
            for (size_t i = 0; i < max_required_buffers; ++i)
```

File: src/cpp/rtps/transport/TCPChannelResourceSecure.cpp [hunk 14]
```diff
@@ -199,8 +199,7 @@ uint32_t TCPChannelResourceSecure::read(
 size_t TCPChannelResourceSecure::send(
         const octet* header,
         size_t header_size,
-        const octet* data,
-        size_t size,
+        const std::array<asio::const_buffer, 3>& send_buffers,
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/TCPChannelResourceSecure.cpp:202]:
```suggestion
        const std::array<asio::const_buffer, max_required_buffers>& send_buffers,
```

File: src/cpp/rtps/transport/TCPChannelResourceSecure.cpp [hunk 15]
```diff
@@ -212,7 +211,13 @@ size_t TCPChannelResourceSecure::send(
         {
             buffers.push_back(asio::buffer(header, header_size));
         }
-        buffers.push_back(asio::buffer(data, size));
+        for (size_t i = 0; i < 3; ++i)
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/TCPChannelResourceSecure.cpp:214]:
```suggestion
        for (size_t i = 0; i < max_required_buffers; ++i)
```

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp [hunk 16]
```diff
@@ -407,13 +413,20 @@ bool test_UDPv4Transport::packet_should_drop(
 }
 
 bool test_UDPv4Transport::log_drop(
-        const octet* buffer,
-        uint32_t size)
+        const std::array<asio::const_buffer, max_required_buffers>& send_buffers,
+        uint32_t total_bytes)
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/test_UDPv4Transport.cpp:417]: There is a `-Wunused-parameter` warning: unused parameter `total_bytes`.

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
I have rerun the tests and the same problems I saw last week are still there. Please, take a look at them to see if they are related (in `master` does not seem to happen as often and consistently as with this PR)

File: src/cpp/rtps/transport/test_UDPv4Transport.cpp (hunk 16 above)
Comment (MiguelCompany) [reply src/cpp/rtps/transport/test_UDPv4Transport.cpp]: Unused parameter removed in a551f3a0c7d0e83d05f6cd6d5d18b8e774cd69c9

**Comment by MiguelCompany** [comment]
> * 242: BlackboxTests_FastRTPS.PubSubFragments.AsyncPubSubAsNonReliableVolatileData300kb.Transport
> * 897: BlackboxTests_DDS_PIM.PubSubFragments.AsyncPubSubAsNonReliableVolatileData300kb.Transport (this one has failed in the Linux CI)

I think these have been solved by 7c930df4d59cbfced843475b216a930cc777b417 

> * 555: BlackboxTests_FastRTPS.BlackBox.TCP_TLS: the following error message appears:
> * 1353: BlackboxTests_DDS_PIM.BlackBox.TCP_TLS: master sometimes blocks and timeouts (rarely). This also happens with this PR (frequently) but the following message is printed that in master does not appear:

I think these have been solved by 4fc658f34714513879d91a688ef15761c67ea39c 

> * 1831: SHMTransportTests.buffer_recover: in master sometimes happens but locally in my machine is failing consistently with this PR:

All the code used by that test has not been modified by this PR...

File: include/fastdds/rtps/messages/RTPSMessageSenderInterface.hpp (hunk 5 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/messages/RTPSMessageSenderInterface.hpp]: Fixed in dc3799349c14b248e1aa6634ca9f354db375f42a

File: src/cpp/rtps/builtin/discovery/participant/DirectMessageSender.hpp [hunk 17]
```diff
@@ -36,54 +36,55 @@ class RTPSParticipantImpl;
...
+         *
+         * @return a const reference to a vector with the GUID of all destinations.
+         */
+        virtual const std::vector<GUID_t>& remote_guids() const override;
+
+        /**
+         * Send a message through this interface.
+         *
+         * @param message Pointer to the buffer with the message already serialized.
+         * @param max_blocking_time_point Future timepoint where blocking send should end.
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/discovery/participant/DirectMessageSender.hpp]: Fixed on 13e3b1fc9a2d2a35347a750cf6a97110bce89599

**Comment by MiguelCompany** [comment]
@JLBuenoLopez-eProsima I think this is finished! 🤞

File: src/cpp/rtps/transport/TCPChannelResourceSecure.cpp [hunk 18]
```diff
@@ -223,8 +228,8 @@ size_t TCPChannelResourceSecure::send(
                 {
                     if (socket->lowest_layer().is_open())
                     {
-                        asio::async_write(*socket, buffers,
-                        [&, socket](const std::error_code& error, const size_t& bytes_transferred)
+                        socket->async_send(buffers,
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/transport/TCPChannelResourceSecure.cpp:231]: This change is giving the following build error with gcc:

```
error: ‘using element_type = class asio::ssl::stream<asio::basic_stream_socket<asio::ip::tcp> >’ {aka ‘class asio::ssl::stream<asio::basic_stream_socket<asio::ip::tcp> >’} has no member named ‘async_send’
  231 |                         socket->async_send(buffers,
```

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

**Comment by JLBuenoLopez-eProsima** [comment]
The following tests are failing:

* 908: BlackboxTests_DDS_PIM.PubSubFragments.AsyncPubSubAsNonReliableVolatileData300kb.Transport: running this test in master 100 times passes successfully, but with these changes the test fails frequently.
* 1812: TCPv4Tests.send_and_receive_between_both_secure_ports: running this test in master prints the following warning and passes successfully:
```
[DEBUG Warning] Error reading TCP header: Operation aborted. -> Function Receive
```
whereas running this PR prints:
```
[DEBUG Warning] Failed to send RTCP message (14 of 19 b): Success -> Function send
[RTCP_MSG_IN Warning] Bad TCP header CRC -> Function Receive
[RTCP_MSG_IN Error] Bad RTCP header identifier, closing connection. -> Function Receive
```
and the test timeouts.
* 1814: TCPv4Tests.send_and_receive_between_secure_clients_1: the error is probably the same as the previous test because the outputs are similar.
* TCP_TLS tests (556 FastRTPS & 1364 DDS PIM) are also finishing with a timeout after printing:
```
[RTCP_MSG_IN Error] Bad RTCP header identifier, closing connection. -> Function Receive
```
It seems the issue could be the same as the previous tests. This does not happen in master.

**Comment by JLBuenoLopez-eProsima** [comment]
Closing this PR. If in the future this feature is revisited, it would probably be easier to start again due to changes in the transport layer.

**Pull request closed without merging**